components/helm-broker/internal/controller/addonsconfiguration_controller.go

```diff
@@ -0,0 +1,500 @@
+		log:    logrus.WithField("controller", "addons-configuration"),
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+
+		chartStorage:  chartStorage,
+		bundleStorage: bundleStorage,
+
+		bundleProvider: bp,
+		protection:     protection{},
+
+		brokerSyncer:      brokerSyncer,
+		brokerFacade:      brokerFacade,
+		docsTopicProvider: docsTopicProvider,
+
+		developMode: developMode,
+		syncBroker:  false,
+	}
+}
+
+// Reconcile reads that state of the cluster for a AddonsConfiguration object and makes changes based on the state read
+// and what is in the AddonsConfiguration.Spec
+func (r *ReconcileAddonsConfiguration) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	instance := &addonsv1alpha1.AddonsConfiguration{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if instance.DeletionTimestamp != nil {
+		if err := r.deleteAddonsProcess(instance); err != nil {
+			return reconcile.Result{RequeueAfter: time.Second * 15}, exerr.Wrapf(err, "while deleting AddonConfiguration %q", request.NamespacedName)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if instance.Status.ObservedGeneration == 0 {
+		r.log.Infof("Start add AddonsConfiguration %s/%s process", instance.Name, instance.Namespace)
+		pendingInstance, err := r.setPendingStatus(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		updatedInstance, err := r.addFinalizer(pendingInstance)
+		if err != nil {
+			return reconcile.Result{Requeue: true}, exerr.Wrapf(err, "while adding a finalizer to AddonsConfiguration %q", request.NamespacedName)
+		}
+		err = r.addAddonsProcess(updatedInstance)
+		if err != nil {
+			return reconcile.Result{}, exerr.Wrapf(err, "while creating AddonsConfiguration %q", request.NamespacedName)
+		}
+		r.log.Info("Add AddonsConfiguration process completed")
+
+	} else if instance.Generation > instance.Status.ObservedGeneration {
+		r.log.Infof("Start update AddonsConfiguration %s/%s process", instance.Name, instance.Namespace)
+
+		instance.Status = addonsv1alpha1.AddonsConfigurationStatus{}
+		err = r.addAddonsProcess(instance)
+		if err != nil {
+			return reconcile.Result{}, exerr.Wrapf(err, "while updating AddonsConfiguration %q", request.NamespacedName)
+		}
+		r.log.Info("Update AddonsConfiguration process completed")
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAddonsConfiguration) addAddonsProcess(addon *addonsv1alpha1.AddonsConfiguration) error {
+	repositories := addons.NewRepositoryCollection()
+
+	r.log.Infof("- load bundles and charts for each addon")
+	for _, specRepository := range addon.Spec.Repositories {
+		if err := specRepository.VerifyURL(r.developMode); err != nil {
+			r.log.Errorf("url %q address is not valid: %s", specRepository.URL, err)
+			continue
+		}
+		r.log.Infof("- create addons for %q repository", specRepository.URL)
+		repo := addons.NewAddonsRepository(specRepository.URL)
```

Comment:
I made mistake here, incorrect URL also should be saved in `addon` status.
Could be something like this:
```
	for _, specRepository := range addon.Spec.Repositories {
		r.log.Infof("- create addons for %q repository", specRepository.URL)
		repo := addons.NewAddonsRepository(specRepository.URL)
		
		if err := specRepository.VerifyURL(r.developMode); err != nil {
			repo.Failed()
			repo.Repository.Reason = addonsv1alpha1.RepositoryURLFetchingError
			repo.Repository.Message = err.Error()
			repositories.AddRepository(repo)

			r.log.Errorf("url %q address is not valid: %s", specRepository.URL, err)
			continue
		}

		addons, err := r.createAddons(specRepository.URL)
		if err != nil {
			repo.Failed()
			repo.Repository.Reason = addonsv1alpha1.RepositoryURLFetchingError
			repo.Repository.Message = err.Error()
			repositories.AddRepository(repo)

			r.log.Errorf("while creating addons for repository from %q: %s", specRepository.URL, err)
			continue
		}

		repo.Addons = addons
		repositories.AddRepository(repo)
	}
```

common fragment could be move to `internal/controller/addons/repository.go`
```
repo.Failed()
repo.Repository.Reason = addonsv1alpha1.RepositoryURLFetchingError
repo.Repository.Message = err.Error()
```